library/src/test/java/com/chuckerteam/chucker/api/ChuckerInterceptorTest.kt

```diff
@@ -566,4 +568,179 @@ internal class ChuckerInterceptorTest {
+
+        val transaction = chuckerInterceptor.expectTransaction()
+
+        assertThat(transaction.requestBody).isEqualTo("olleH")
+        assertThat(transaction.responseBody).isEqualTo("eybdooG")
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = ClientFactory::class)
+    fun nextBodyDecoder_isUsed_whenPreviousDoesNotDecode(factory: ClientFactory) {
+        val chuckerInterceptor = ChuckerInterceptorDelegate(
+            cacheDirectoryProvider = { tempDir },
+            decoders = listOf(NoOpDecoder(), LiteralBodyDecoder()),
+        )
+        val client = factory.create(chuckerInterceptor)
+        val request = Request.Builder().url(serverUrl)
+            .post("Hello".toRequestBody())
+            .build()
+        server.enqueue(MockResponse().setBody("Goodbye"))
+
+        client.newCall(request).execute().readByteStringBody()
+
+        val transaction = chuckerInterceptor.expectTransaction()
+
+        assertThat(transaction.requestBody).isEqualTo("Request")
+        assertThat(transaction.responseBody).isEqualTo("Response")
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = ClientFactory::class)
+    fun bodyDecoder_canThrowIoExceptions(factory: ClientFactory) {
+        val chuckerInterceptor = ChuckerInterceptorDelegate(
+            cacheDirectoryProvider = { tempDir },
+            decoders = listOf(IoThrowingDecoder(), LiteralBodyDecoder()),
+        )
+        val client = factory.create(chuckerInterceptor)
+        val request = Request.Builder().url(serverUrl)
+            .post("Hello".toRequestBody())
+            .build()
+        server.enqueue(MockResponse().setBody("Goodbye"))
+
+        client.newCall(request).execute().readByteStringBody()
+
+        val transaction = chuckerInterceptor.expectTransaction()
+
+        assertThat(transaction.requestBody).isEqualTo("Request")
+        assertThat(transaction.responseBody).isEqualTo("Response")
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = ClientFactory::class)
+    fun bodyDecoders_areAppliedLazily(factory: ClientFactory) {
+        val statefulDecoder = StatefulDecoder()
+        val chuckerInterceptor = ChuckerInterceptorDelegate(
+            cacheDirectoryProvider = { tempDir },
+            decoders = listOf(LiteralBodyDecoder(), statefulDecoder),
+        )
+        val client = factory.create(chuckerInterceptor)
+        val request = Request.Builder().url(serverUrl)
+            .post("Hello".toRequestBody())
+            .build()
+        server.enqueue(MockResponse().setBody("Goodbye"))
+
+        client.newCall(request).execute().readByteStringBody()
+
+        assertThat(statefulDecoder.didDecodeRequest).isFalse()
+        assertThat(statefulDecoder.didDecodeResponse).isFalse()
+    }
+
+    private class LiteralBodyDecoder : BodyDecoder {
+        override fun decodeRequest(request: Request, body: ByteString) = "Request"
+        override fun decodeResponse(response: Response, body: ByteString) = "Response"
+    }
```

Comment:
I believe that all the `*Decoder` + all the related tests should be moved to a `ChuckerInterceptor_withDecoderTest`